mozart++/system/process.hpp

```diff
@@ -0,0 +1,504 @@
+            if (!redirect_or_pipe(startup._stderr, pstderr)) {
+                close_pipe(pstdin);
+                close_pipe(pstdout);
+                mpp::throw_ex<mpp::runtime_error>("unable to bind stderr");
+            }
+        }
+
+        create_process_impl(startup, info, pstdin, pstdout, pstderr);
+    }
+
+    void close_process(process_info &info) {
+#ifdef _WIN32
+        mpp_impl::close_fd(info._pid);
+        mpp_impl::close_fd(info._tid);
+#endif
+        mpp_impl::close_fd(info._stdin);
+        mpp_impl::close_fd(info._stdout);
+        mpp_impl::close_fd(info._stderr);
+    }
+
+    int wait_for(const process_info &info) {
+#ifdef _WIN32
+        WaitForSingleObject(info._pid, INFINITE);
+        DWORD code = 0;
+        GetExitCodeProcess(info._pid, &code);
+        return code;
+#else
+        int status;
+        if (waitpid(info._pid, &status, 0) == info._pid) {
+            if (WIFEXITED(status)) {
+                return WEXITSTATUS(status);
+            }
+
+            if (WIFSIGNALED(status)) {
+                return WTERMSIG(status);
+            }
+
+            if (WIFSTOPPED(status)) {
+                return WSTOPSIG(status);
+            }
+        }
+        return 0;
+#endif
+    }
+
+    void terminate_process(const process_info &info, bool force) {
+#ifdef _WIN32
+        TerminateProcess(info._pid, 0);
+#else
+        kill(info._pid, force ? SIGKILL : SIGTERM);
+#endif
+    }
+}
+
+namespace mpp {
+    using mpp_impl::redirect_info;
+    using mpp_impl::process_info;
+    using mpp_impl::process_startup;
+    using mpp_impl::fd_type;
+
+    class process {
+        friend class process_builder;
+
+    private:
+        process_info _info;
+        int _exit_code = -1;
+        std::unique_ptr<fdostream> _stdin;
+        std::unique_ptr<fdistream> _stdout;
+        std::unique_ptr<fdistream> _stderr;
+
+        explicit process(const process_info &info)
+            : _info(info),
+              _stdin(std::make_unique<fdostream>(_info._stdin)),
+              _stdout(std::make_unique<fdistream>(_info._stdout)),
+              _stderr(std::make_unique<fdistream>(_info._stderr)) {
+        }
+
+    public:
+        process() = delete;
+
+        process(const process &) = delete;
+
+        process(process &&) = default;
+
+        process &operator=(process &&) = delete;
+
+        process &operator=(const process &) = delete;
+
+    public:
+        ~process() {
+            mpp_impl::close_process(_info);
+        }
+
+        std::ostream &in() {
+            return *_stdin;
+        }
+
+        std::istream &out() {
+            return *_stdout;
+        }
+
+        std::istream &err() {
+            return *_stderr;
+        }
+
+        int wait_for() {
+            if (_exit_code >= 0) {
+                return _exit_code;
+            }
+            _exit_code = mpp_impl::wait_for(_info);
+            return _exit_code;
+        }
+
+        bool is_exited() const {
```

Comment:
I think this is not very reasonable. If user use this function but don't use wait_for, the exit code will never change.

Reply:
See new approach